### PR TITLE
SONARJAVA-1906 Rule S3067: "getClass" should not be used for synchron…

### PR DESCRIPTION
**Pull request by saberduck**
Implements Rule S3067

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/synchronization/SynchronizationOnGetClassCheck.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      if (synchronizedTree.expression().is(METHOD_INVOCATION)) {
+        MethodInvocationTree synchronizedExpr = (MethodInvocationTree) synchronizedTree.expression();
+        if (GET_CLASS_MATCHER.matches(synchronizedExpr) && !isEnclosingClassFinal(synchronizedExpr.methodSelect())) {
+          reportIssue(synchronizedExpr, "Synchronize on the static class name instead.");
+        }
+      }
+    }
+  }
+
+  private boolean isEnclosingClassFinal(ExpressionTree expressionTree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/synchronization/SynchronizationOnGetClassCheck.java]: MAJOR Make "isEnclosingClassFinal" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/synchronization/SynchronizationOnGetClassCheck.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  private static boolean isEnclosingClassFinal(ExpressionTree expressionTree) {
+    //if it's identifier we go up the tree to get the method and check if owner of the method is final
+    if (expressionTree.is(IDENTIFIER)) {
+      Tree parent = expressionTree.parent();
+      while (!parent.is(METHOD)) {
+        parent = parent.parent();
+      }
+      return ((MethodTree) parent).symbol().owner().isFinal();
+    } else {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/synchronization/SynchronizationOnGetClassCheck.java]: I would remove the else given that you return at the end of the if.

File: java-checks/src/test/files/checks/SynchronizationOnGetClassCheck.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+public class MyClass {
+  public void doSomethingSynchronized() {
+    synchronized (getClass()) { // Noncompliant [[sc=19;ec=29]] {{Synchronize on the static class name instead.}}
+      // ...
```
Comment (benzonico) [inline java-checks/src/test/files/checks/SynchronizationOnGetClassCheck.java]: No need for that comment... (and the others... ;) )

File: java-checks/src/test/files/checks/SynchronizationOnGetClassCheck.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    }
+  }
+
+  public void doSomethingSynchronized2() {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/SynchronizationOnGetClassCheck.java]: Given the name of that method there's probably no need to put this second issue in a different method than the first one.

File: java-checks/src/test/files/checks/SynchronizationOnGetClassCheck.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+public final class FinalClassIsCompliant {
+  public void doSomethingSynchronized() {
+    synchronized (this.getClass()) { // Compliant
+      // ...
+    }
+  }
+}
+
+public class MemberSelect {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/SynchronizationOnGetClassCheck.java]: No need for a separate class, named the method appropriately and put it as a member of the first class ?

**Pull request merged**